spices/SPICE-0024-annotation-converters.adoc

```diff
@@ -0,0 +1,405 @@
+  requiredProperty: String
+
+  @Available { since = 2 }
+  addedInV2: String?
+
+  children: Listing<NestedConfig>
+}
+
+class Available extends Annotation {
+  since: Int?
+  until: Int?
+
+  function validate(target: Int, value: Any): Any =
+    if (value == null)
+      null
+    else if (since != null && target < since)
+      throw("Target version \(target) is lower than minimum version \(since)")
+    else if (until != null && until < target)
+      throw("Target version \(target) is higher than maximum version \(until)")
+    else
+      value
+}
+
+output {
+  renderer {
+    annotationConverters {
+      [Available] = (propertyName, annotation, value) ->
+        Pair(propertyName, annotation.validate(softwareVersion, value))
+    }
+  }
+}
+----
+
+=== Language Binding Libraries
+
+While annotation converters do not directly affect Pkl's language bindings, it is worth highlighting how they _can_ interact.
+The "Validating properties in nested structures" example above shows how annotation converters may be used for validations that need data "injected" from the top-level module being rendered.
+Most usage of Pkl's language bindings bypass renderers (and therefore converters of all kinds) entirely, which would skip such validation.
+
+Clients of Pkl's language bindings can change how they evaluate modules to take advantage of annotation converters.
+This is done by combining a small amount of Pkl module code with some changes on the client side.
+
+In Pkl, set up the link:{uri-bindings-specification}[`pkl-binary`] renderer:
+
+[source,pkl]
+----
+import "pkl:pklbinary"
+
+output {
+  renderer = new pklbinary.Renderer {
+    annotationConverters {
+    // Add validating converters here!
+    }
+  }
+}
+----
+
+In the client code, update how modules are evaluated:
+
+|===
+|Language |Old invocation |New invocation
+
+|Java/Kotlin
+|`ConfigEvaluator.evaluate(moduleSource)`
+|`Evaluator.evaluateOutputBytes(moduleSource)`
+
+`Config.fromPklBinary(bytes)`
+
+|Go
+|`Evaluator.EvaluateModule(ctx, moduleSource, &out)`
+|`Evaluator.EvaluateOutputBytes(ctx, moduleSource)`
+
+`pkl.Unmarshal(buf)`
+
+|Swift
+|`Evaluator.evaluateModule(moduleSource, T.self)`
+|`Evaluator.evaluateOutputBytes(moduleSource)`
+
+`PklDecoder.decode(T.self, from: bytes)`
+
+|===
+
+== Compatibility
+
+This is a purely additive change.
+Modules that adopt annotation converters will not be compatible with prior versions of Pkl.
+
+== Future directions
+
+=== Built-in annotations for built-in renderers
+
+One possible extension to annotation converters might be built-in annotations for specific renderers.
+These annotation converters could be configured by default on their respective renderers, allowing per-format customization for simple cases without needing to configure any converters directly:
+
+[source,pkl]
+----
+// fmt.pkl
+import "pkl:json"
+import "pkl:yaml"
+
+@json.Property { name = "foo_bar" }
+@yaml.Property { name = "foo-bar" }
+fooBar: String = "hello world"
```

Comment:
I actually think we should add these two as part of the spice; `json.Property` and `yaml.Property` both make a lot of sense. I think we can add ones for xml, protobuf and jsonnet pretty easily too.

Properties, PCF and Plist are a little trickier, though, because they don't have their own modules and I'd prefer to avoid adding this to `base.pkl`. That doesn't need to be a blocker for the existing annotations, though.